### ISSUE TITLE
请问关于测试代码是否有样例演示

### ISSUE THREAD

**Issue by youflyaway** [TUDB-Labs/MoE-PEFT#18]
非常感谢您的工作！

请问这个项目该如何评估结果，我训练完数据集后推理出来的结果有很多幻觉，我该如何加载训练后的权重

**Comment by mikecovlee** [comment]
Just run:
```bash
./launch.py run <PATH TO YOUR MODEL> --evaluate
```

**Comment by mikecovlee** [comment]
你可以详细描述一下你的需求，是需要推理、评估还是继续训练呢？

**Comment by youflyaway** [comment]
需要推理和对结果进行评估

**Comment by mikecovlee** [comment]
自动评估
```bash
./launch.py run <PATH TO YOUR MODEL> --evaluate
```
推理
```bash
./launch.py run <PATH TO YOUR MODEL> --inference
```
有其他不太清楚的，可以看下help
```bash
./launch.py help
```

**Comment by youflyaway** [comment]
> 自动评估
> 
> ```shell
> ./launch.py run <PATH TO YOUR MODEL> --evaluate
> ```
> 
> 推理
> 
> ```shell
> ./launch.py run <PATH TO YOUR MODEL> --inference
> ```
> 
> 有其他不太清楚的，可以看下help
> 
> ```shell
> ./launch.py help
> ```

我训练完进行推理，为什么有太多的幻觉，我训练的时候只让模型输出几个词，但是推理的时候会输出长篇大论给我，这个是我没训练好还是参数没调对

```
{
    "cutoff_len": 2048,
    "save_step": 1000,
    "train_lora_candidate_num": 2,
    "train_lora_simultaneously_num": 2,
    "train_strategy": "optim",
    "lora": [
        {
            "name": "casual_0",
            "task_name": "casual",
            "optim": "adamw",
            "scheduler_type": "constant",
            "warmup_steps": 0,
            "lr": 0.00005,
            "batch_size": 1,
            "micro_batch_size": 1,
            "evaluate_batch_size": 1,
            "num_epochs": 2,
            "r": 8,
            "lora_alpha": 16,
            "lora_dropout": 0.1,
            "target_modules": {
                "q_proj": true,
                "k_proj": true,
                "v_proj": true,
                "o_proj": true,
                "gate_proj": true,
                "down_proj": true,
                "up_proj": true
            },
            "routing_strategy": "mixlora",
            "num_experts": 4,
            "group_by_length": false,
            "data": "/home/cy/code/workplace/llm/dataset/alpaca",
            "prompt": "alpaca"
        }
    ]
}
```
这是我的参数设置

**Comment by swf-cqu** [comment]
你好，关于代码，我发现训练时，训练数据并不会random，在每个epoch都会按照顺序的batch去训练模型，这是否会对最终的训练结果有影响？

**Comment by mikecovlee** [comment]
> 你好，关于代码，我发现训练时，训练数据并不会random，在每个epoch都会按照顺序的batch去训练模型，这是否会对最终的训练结果有影响？

这是为了保证可复现性。如果你需要更多的随机性，可调整 random seed

**Comment by mikecovlee** [comment]
> > 自动评估
> > ```shell
> > ./launch.py run <PATH TO YOUR MODEL> --evaluate
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 推理
> > ```shell
> > ./launch.py run <PATH TO YOUR MODEL> --inference
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 有其他不太清楚的，可以看下help
> > ```shell
> > ./launch.py help
> > ```
> 
> 我训练完进行推理，为什么有太多的幻觉，我训练的时候只让模型输出几个词，但是推理的时候会输出长篇大论给我，这个是我没训练好还是参数没调对
> 
> ```
> {
>     "cutoff_len": 2048,
>     "save_step": 1000,
>     "train_lora_candidate_num": 2,
>     "train_lora_simultaneously_num": 2,
>     "train_strategy": "optim",
>     "lora": [
>         {
>             "name": "casual_0",
>             "task_name": "casual",
>             "optim": "adamw",
>             "scheduler_type": "constant",
>             "warmup_steps": 0,
>             "lr": 0.00005,
>             "batch_size": 1,
>             "micro_batch_size": 1,
>             "evaluate_batch_size": 1,
>             "num_epochs": 2,
>             "r": 8,
>             "lora_alpha": 16,
>             "lora_dropout": 0.1,
>             "target_modules": {
>                 "q_proj": true,
>                 "k_proj": true,
>                 "v_proj": true,
>                 "o_proj": true,
>                 "gate_proj": true,
>                 "down_proj": true,
>                 "up_proj": true
>             },
>             "routing_strategy": "mixlora",
>             "num_experts": 4,
>             "group_by_length": false,
>             "data": "/home/cy/code/workplace/llm/dataset/alpaca",
>             "prompt": "alpaca"
>         }
>     ]
> }
> ```
> 
> 这是我的参数设置

幻觉很正常，原因很复杂，仅根据你给的这些信息我没办法判断，具体还是要多调调超参。SFT 本质上还是依赖 base model 的能力

**Comment by youflyaway** [comment]
> > > 自动评估
> > > ```shell
> > > ./launch.py run <PATH TO YOUR MODEL> --evaluate
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 推理
> > > ```shell
> > > ./launch.py run <PATH TO YOUR MODEL> --inference
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 有其他不太清楚的，可以看下help
> > > ```shell
> > > ./launch.py help
> > > ```
> > 
> > 
> > 我训练完进行推理，为什么有太多的幻觉，我训练的时候只让模型输出几个词，但是推理的时候会输出长篇大论给我，这个是我没训练好还是参数没调对
> > ```
> > {
> >     "cutoff_len": 2048,
> >     "save_step": 1000,
> >     "train_lora_candidate_num": 2,
> >     "train_lora_simultaneously_num": 2,
> >     "train_strategy": "optim",
> >     "lora": [
> >         {
> >             "name": "casual_0",
> >             "task_name": "casual",
> >             "optim": "adamw",
> >             "scheduler_type": "constant",
> >             "warmup_steps": 0,
> >             "lr": 0.00005,
> >             "batch_size": 1,
> >             "micro_batch_size": 1,
> >             "evaluate_batch_size": 1,
> >             "num_epochs": 2,
> >             "r": 8,
> >             "lora_alpha": 16,
> >             "lora_dropout": 0.1,
> >             "target_modules": {
> >                 "q_proj": true,
> >                 "k_proj": true,
> >                 "v_proj": true,
> >                 "o_proj": true,
> >                 "gate_proj": true,
> >                 "down_proj": true,
> >                 "up_proj": true
> >             },
> >             "routing_strategy": "mixlora",
> >             "num_experts": 4,
> >             "group_by_length": false,
> >             "data": "/home/cy/code/workplace/llm/dataset/alpaca",
> >             "prompt": "alpaca"
> >         }
> >     ]
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 这是我的参数设置
> 
> 幻觉很正常，原因很复杂，仅根据你给的这些信息我没办法判断，具体还是要多调调超参。SFT 本质上还是依赖 base model 的能力

感谢您的回复！🙏

**Comment by swf-cqu** [comment]
> > 你好，关于代码，我发现训练时，训练数据并不会random，在每个epoch都会按照顺序的batch去训练模型，这是否会对最终的训练结果有影响？
> 
> 这是为了保证可复现性。如果你需要更多的随机性，可调整 random seed

感谢您的回复，我还有一个问题，想要请教您，我发现模型在推理时，尽管默认设置的batch_size 为8，但是实际还是一个样本一个样本的推理，我试图设置concurrent_jobs 伟8，但依旧是逐个样本推理，请问您有什么好的建议吗？感谢您的指导

**Comment by mikecovlee** [comment]
> > > 你好，关于代码，我发现训练时，训练数据并不会random，在每个epoch都会按照顺序的batch去训练模型，这是否会对最终的训练结果有影响？
> > 
> > 
> > 这是为了保证可复现性。如果你需要更多的随机性，可调整 random seed
> 
> 感谢您的回复，我还有一个问题，想要请教您，我发现模型在推理时，尽管默认设置的batch_size 为8，但是实际还是一个样本一个样本的推理，我试图设置concurrent_jobs 伟8，但依旧是逐个样本推理，请问您有什么好的建议吗？感谢您的指导

你仔细看看呢，肯定是并行的。我刚测试过了，concurrent_jobs设置上去就 OK 了

**Comment by swf-cqu** [comment]
> > > > 你好，关于代码，我发现训练时，训练数据并不会random，在每个epoch都会按照顺序的batch去训练模型，这是否会对最终的训练结果有影响？
> > > 
> > > 
> > > 这是为了保证可复现性。如果你需要更多的随机性，可调整 random seed
> > 
> > 
> > 感谢您的回复，我还有一个问题，想要请教您，我发现模型在推理时，尽管默认设置的batch_size 为8，但是实际还是一个样本一个样本的推理，我试图设置concurrent_jobs 伟8，但依旧是逐个样本推理，请问您有什么好的建议吗？感谢您的指导
> 
> 你仔细看看呢，肯定是并行的。我刚测试过了，concurrent_jobs设置上去就 OK 了
def _batch_generate(
    model: LLMModel,
    tokenizer: Tokenizer,
    max_gen_len: Optional[int],
    use_cache: bool,
    cache_implementation: Optional[str],
    stream_callback: Optional[Callable],
    config_dict: Dict[str, GenerateConfig],
    current_jobs: List[GenerateData],
    batch_config: List[LLMBatchConfig],
    input_tokens: List[Tokens],
    max_tokens_len: int,
    min_tokens_len: int,
):
    executor.empty_cache()
    device = torch.device(model.device_)
    batch_size = len(input_tokens)
    if max_gen_len is None:
        max_gen_len = model.config_.max_seq_len_ - max_tokens_len
    total_len = min(model.config_.max_seq_len_, max_gen_len + max_tokens_len)

    past_key_values = (
        cache_factory(
            cache_implementation=cache_implementation,
            config=model.model_.model_config(),
            batch_size=batch_size,
            max_cache_len=total_len,
        )
        if cache_implementation is not None
        else None
    )

    tokens = torch.full(
        (batch_size, total_len), tokenizer.pad_id_, dtype=torch.int64, device=device
    )
    for k, t in enumerate(input_tokens):
        tokens[k, : len(t)] = torch.tensor(t, dtype=torch.int64, device=device)

    prev_pos = 0
    stop_reached = torch.tensor([False] * batch_size, device=device)
    input_text_mask = tokens != tokenizer.pad_id_
    for cur_pos in range(min_tokens_len, total_len):
        input_data = LLMModelInput(
            batch_configs_=batch_config,
            batch_tokens_=tokens[:, prev_pos:cur_pos].tolist(),
            inference_mode_=True,
        )
        outputs = model.forward(input_data, past_key_values)
        for output in outputs:
            config = config_dict[output.adapter_name]
            start_idx = output.batch_start_idx_
            end_idx = output.batch_end_idx_

            next_token = logits_process(
                output.logits[:, -1],
                tokens[start_idx:end_idx, :cur_pos],
                config.temperature,
                config.top_p,
                config.top_k,
                config.do_sample,
                config.repetition_penalty,
                config.renormalize_logits,
            )

            next_token = torch.where(
                input_text_mask[start_idx:end_idx, cur_pos],
                tokens[start_idx:end_idx, cur_pos],
                next_token,
            ).to(torch.int64)
            
            tokens[start_idx:end_idx, cur_pos] = next_token
            stop_criteria = (~input_text_mask[start_idx:end_idx, cur_pos]) & (
                next_token == config.stop_token_
            )
            stop_reached[start_idx:end_idx] |= stop_criteria

        stop_reached |= total_len - cur_pos == 1

        if any(stop_reached):
            break

        if stream_callback is not None:
            stream_callback(
                cur_pos,
                _gen_outputs(
                    tokenizer,
                    config_dict,
                    current_jobs,
                    tokens,
                ),
            )

        if use_cache:
            prev_pos = cur_pos

    return _dispatch_task_out(
        tokenizer, config_dict, current_jobs, tokens, stop_reached
    )
    
    我仔细查看了相关代码，我运行的是generate.py文件，我发现，这个文件最终会运行以上的代码，在这个代码中， 存在以下的条件句：    
       if any(stop_reached):
            break
            这段代码，我理解的是只要有一个sample满足停止条件，就会跳出循环，其余的样本又会添加到config.中的data中（详见moe_peft.generator.py文件中的如下代码：
               for data in running_jobs:
                       config_dict[data.adapter_name_].data_.append(data)
 
 所以我理解的是，尽管是一个批次的样本在进行推理，但是实际只要有一个sample达到终止条件，就会停止生成，最终的生成效果实际是一个样本一个样本在生成？不知道我理解的是否有误，期待您的回复
